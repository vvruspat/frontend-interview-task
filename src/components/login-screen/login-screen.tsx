import React, {
  ChangeEvent,
  FormEvent,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import { useAppDispatch } from "../../store/hooks";
import { login, selectLoginStatus } from "../../store/login/loginSlice";
import { EFetchingStatus } from "../../types/fetching-status.type";
import {
  Error,
  Logo,
  Label,
  Input,
  Heading,
  Form,
  FormRow,
  PrimaryButton,
  Password,
} from "../kit";
import { LoginScreenRoot } from "./login-screen.styled";

export const LoginScreen = () => {
  const loginStatus = useSelector(selectLoginStatus);
  const history = useHistory();

  const [password, setPassword] = useState("");
  const [username, setUsername] = useState("");

  const dispatch = useAppDispatch();

  const disabled = useMemo(() => {
    return !password || !username || loginStatus === EFetchingStatus.FETCHING;
  }, [password, username, loginStatus]);

  const onSubmit = useCallback(
    (e: FormEvent) => {
      e.preventDefault();
      dispatch(login(username, password));
    },
    [dispatch, username, password],
  );

  useEffect(() => {
    if (loginStatus === EFetchingStatus.SUCCESS) {
      history.push("/");
    }
  }, [loginStatus, history]);

  return (
    <LoginScreenRoot>
      <Logo />
      <section>
        <Heading>Log in</Heading>
        <Form onSubmit={onSubmit}>
          <FormRow>
            <Label htmlFor={"login"}>Username</Label>
            <Input
              type={"text"}
              placeholder={"login"}
              id={"login"}
              data-testid={"login-input"}
              onChange={(e: ChangeEvent<HTMLInputElement>) =>
                setUsername(e.target.value)
              }
            />
          </FormRow>
          <FormRow>
            <Label htmlFor={"password"}>Password</Label>
            <Password
              id={"password"}
              placeholder={"password"}
              data-testid={"password-input"}
              onChange={(e: ChangeEvent<HTMLInputElement>) =>
                setPassword(e.target.value)
              }
            />
          </FormRow>
          {loginStatus === EFetchingStatus.ERROR && (
            <Error size="medium">Invalid username or password</Error>
          )}
          <FormRow>
            <PrimaryButton
              type={"submit"}
              disabled={disabled}
              data-testid={"submit-button"}
            >
              Log in
            </PrimaryButton>
          </FormRow>
        </Form>
      </section>
    </LoginScreenRoot>
  );
};
