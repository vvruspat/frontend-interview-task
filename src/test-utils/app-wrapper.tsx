import React, { FC } from "react";
import { Provider } from "react-redux";
import { GlobalStyles } from "../components/global-styles";
import { AuthProvider } from "../auth";
import { store } from "../store/store";

export const AppWrapper: FC = ({ children }) => {
  return (
    <Provider store={store}>
      <GlobalStyles />
      <AuthProvider>{children}</AuthProvider>
    </Provider>
  );
};
