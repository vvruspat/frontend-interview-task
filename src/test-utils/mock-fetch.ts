import { error401Data } from "./test-data/error";
import { feedbacksData } from "./test-data/feedbacks";
import { jwtData } from "./test-data/jwt";
import { themesData } from "./test-data/themes";

export function mockFetch() {
  let status: number = 200;

  return {
    setStatus: (newStatus: number) => {
      status = newStatus;
    },

    fetch: jest.fn().mockImplementation((url: string) => {
      if (status !== 200) {
        return Promise.reject(error401Data);
      }

      if (url.indexOf("/api/reviews") !== -1) {
        return Promise.resolve({
          ok: true,
          status: status,
          json: () => Promise.resolve({ data: feedbacksData }),
        });
      }

      if (url.indexOf("/api/themes") !== -1) {
        return Promise.resolve({
          ok: true,
          status: status,
          json: () => Promise.resolve({ data: themesData }),
        });
      }

      if (url.indexOf("/login") !== -1) {
        return Promise.resolve({
          ok: true,
          status: status,
          json: () => Promise.resolve(jwtData),
        });
      }

      return Promise.resolve({
        ok: true,
        status: status,
        json: () => {},
      });
    }),
  };
}
