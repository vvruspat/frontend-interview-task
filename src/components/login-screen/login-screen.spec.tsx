import { createMemoryHistory } from "history";
import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { MemoryRouter, Router } from "react-router-dom";
import { AppWrapper } from "../../test-utils/app-wrapper";
import { mockFetch } from "../../test-utils/mock-fetch";
import { LoginScreen } from "./login-screen";

describe("LoginScreen", () => {
  test("renders login", async () => {
    const mock = mockFetch();
    window.fetch = mock.fetch;

    const history = createMemoryHistory();

    const component = render(
      <MemoryRouter>
        <Router history={history}>
          <LoginScreen />
        </Router>
      </MemoryRouter>,
      { wrapper: AppWrapper },
    );

    const loginInput = await component.findByTestId("login-input");
    const passwordInput = await component.findByTestId("password-input");
    const submitButton = await component.findByTestId("submit-button");

    expect(component).toMatchSnapshot();

    expect(loginInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toBeDisabled();

    fireEvent.change(loginInput, { target: { value: "testusername" } });

    expect(submitButton).toBeDisabled();

    fireEvent.change(passwordInput, { target: { value: "testpassword" } });

    expect(submitButton).not.toBeDisabled();

    expect(component).toMatchSnapshot();

    expect(history.length).toBe(1);

    fireEvent.click(submitButton);

    await new Promise((resolve) => setTimeout(resolve, 100));

    expect(history.length).toBe(2);
  });

  test("renders login failed 401", async () => {
    const mock = mockFetch();
    window.fetch = mock.fetch;

    mock.setStatus(401);

    const history = createMemoryHistory();

    const component = render(
      <MemoryRouter>
        <Router history={history}>
          <LoginScreen />
        </Router>
      </MemoryRouter>,
      { wrapper: AppWrapper },
    );

    const loginInput = await component.findByTestId("login-input");
    const passwordInput = await component.findByTestId("password-input");
    const submitButton = await component.findByTestId("submit-button");

    fireEvent.change(loginInput, { target: { value: "testusername" } });
    fireEvent.change(passwordInput, { target: { value: "testpassword" } });

    fireEvent.click(submitButton);

    await new Promise((resolve) => setTimeout(resolve, 100));

    const errorMessage = await component.findByText(
      "Invalid username or password",
    );

    expect(errorMessage).toBeVisible();
  });
});
