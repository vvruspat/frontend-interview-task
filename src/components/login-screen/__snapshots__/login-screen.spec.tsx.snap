// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders login 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-kfzAmx jflWSF"
      >
        <div
          class="sc-kEjbxe kemURr"
        />
        <section>
          <h1
            class="sc-fubCfw jnQpQs"
          >
            Log in
          </h1>
          <form
            class="sc-gKsewC bqPZIs"
          >
            <div
              class="sc-iBPRYJ kEQcfl"
            >
              <label
                class="sc-jrAGrp edkgue"
                for="login"
              >
                Username
              </label>
              <input
                class="sc-pFZIQ hDHmTa"
                data-testid="login-input"
                id="login"
                placeholder="login"
                type="text"
              />
            </div>
            <div
              class="sc-iBPRYJ kEQcfl"
            >
              <label
                class="sc-jrAGrp edkgue"
                for="password"
              >
                Password
              </label>
              <div
                class="sc-iqHYGH gfENmc"
              >
                <input
                  class="sc-pFZIQ sc-crrsfI hDHmTa gRmKNV"
                  data-testid="password-input"
                  id="password"
                  placeholder="password"
                  type="password"
                />
                <button
                  class="sc-dQppl sc-bqyKva fRMbOY bNERPa"
                  data-testid="hide-password-button"
                />
              </div>
            </div>
            <div
              class="sc-iBPRYJ kEQcfl"
            >
              <button
                class="sc-bdfBwQ sc-gsTCUz ddILAN fxcCaX"
                data-testid="submit-button"
                disabled=""
                type="submit"
              >
                Log in
              </button>
            </div>
          </form>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-kfzAmx jflWSF"
    >
      <div
        class="sc-kEjbxe kemURr"
      />
      <section>
        <h1
          class="sc-fubCfw jnQpQs"
        >
          Log in
        </h1>
        <form
          class="sc-gKsewC bqPZIs"
        >
          <div
            class="sc-iBPRYJ kEQcfl"
          >
            <label
              class="sc-jrAGrp edkgue"
              for="login"
            >
              Username
            </label>
            <input
              class="sc-pFZIQ hDHmTa"
              data-testid="login-input"
              id="login"
              placeholder="login"
              type="text"
            />
          </div>
          <div
            class="sc-iBPRYJ kEQcfl"
          >
            <label
              class="sc-jrAGrp edkgue"
              for="password"
            >
              Password
            </label>
            <div
              class="sc-iqHYGH gfENmc"
            >
              <input
                class="sc-pFZIQ sc-crrsfI hDHmTa gRmKNV"
                data-testid="password-input"
                id="password"
                placeholder="password"
                type="password"
              />
              <button
                class="sc-dQppl sc-bqyKva fRMbOY bNERPa"
                data-testid="hide-password-button"
              />
            </div>
          </div>
          <div
            class="sc-iBPRYJ kEQcfl"
          >
            <button
              class="sc-bdfBwQ sc-gsTCUz ddILAN fxcCaX"
              data-testid="submit-button"
              disabled=""
              type="submit"
            >
              Log in
            </button>
          </div>
        </form>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LoginScreen renders login 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-kfzAmx jflWSF"
      >
        <div
          class="sc-kEjbxe kemURr"
        />
        <section>
          <h1
            class="sc-fubCfw jnQpQs"
          >
            Log in
          </h1>
          <form
            class="sc-gKsewC bqPZIs"
          >
            <div
              class="sc-iBPRYJ kEQcfl"
            >
              <label
                class="sc-jrAGrp edkgue"
                for="login"
              >
                Username
              </label>
              <input
                class="sc-pFZIQ hDHmTa"
                data-testid="login-input"
                id="login"
                placeholder="login"
                type="text"
              />
            </div>
            <div
              class="sc-iBPRYJ kEQcfl"
            >
              <label
                class="sc-jrAGrp edkgue"
                for="password"
              >
                Password
              </label>
              <div
                class="sc-iqHYGH gfENmc"
              >
                <input
                  class="sc-pFZIQ sc-crrsfI hDHmTa gRmKNV"
                  data-testid="password-input"
                  id="password"
                  placeholder="password"
                  type="password"
                />
                <button
                  class="sc-dQppl sc-bqyKva fRMbOY bNERPa"
                  data-testid="hide-password-button"
                />
              </div>
            </div>
            <div
              class="sc-iBPRYJ kEQcfl"
            >
              <button
                class="sc-bdfBwQ sc-gsTCUz ddILAN fxcCaX"
                data-testid="submit-button"
                type="submit"
              >
                Log in
              </button>
            </div>
          </form>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-kfzAmx jflWSF"
    >
      <div
        class="sc-kEjbxe kemURr"
      />
      <section>
        <h1
          class="sc-fubCfw jnQpQs"
        >
          Log in
        </h1>
        <form
          class="sc-gKsewC bqPZIs"
        >
          <div
            class="sc-iBPRYJ kEQcfl"
          >
            <label
              class="sc-jrAGrp edkgue"
              for="login"
            >
              Username
            </label>
            <input
              class="sc-pFZIQ hDHmTa"
              data-testid="login-input"
              id="login"
              placeholder="login"
              type="text"
            />
          </div>
          <div
            class="sc-iBPRYJ kEQcfl"
          >
            <label
              class="sc-jrAGrp edkgue"
              for="password"
            >
              Password
            </label>
            <div
              class="sc-iqHYGH gfENmc"
            >
              <input
                class="sc-pFZIQ sc-crrsfI hDHmTa gRmKNV"
                data-testid="password-input"
                id="password"
                placeholder="password"
                type="password"
              />
              <button
                class="sc-dQppl sc-bqyKva fRMbOY bNERPa"
                data-testid="hide-password-button"
              />
            </div>
          </div>
          <div
            class="sc-iBPRYJ kEQcfl"
          >
            <button
              class="sc-bdfBwQ sc-gsTCUz ddILAN fxcCaX"
              data-testid="submit-button"
              type="submit"
            >
              Log in
            </button>
          </div>
        </form>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
